tools/src/h5dump/h5dump_ddl.c

```diff
@@ -455,7 +457,13 @@ dump_all_cb(hid_t group, const char *name, const H5L_info2_t *linfo, void H5_ATT
                 break;
 
             case H5L_TYPE_EXTERNAL:
-                if ((targbuf = (char *)malloc(linfo->u.val_size)) == NULL) {
+                if (H5Lget_info2(group, name, &local_linfo, H5P_DEFAULT) < 0) {
```

Comment:
No particular problem with these changes; I'll just bring up again that I think this points to a fundamental design problem that must be fixed and these changes are hiding that. I believe it's reasonable that if you write some tool/code to iterate over links in a group, you should be able to assume that using the info in the `H5L_info2_t` that is passed to your callback will always be safe. By having to call `H5Lget_info2` here, you're essentially doing the same work twice and this may contribute to more performance problems in link iteration. I know that there are applications out there using this exact design pattern, so we will continue to hit this issue.